Modules/posixmodule.c

```diff
@@ -1523,11 +1523,12 @@ win32_wchdir(LPCWSTR path)
             SetLastError(ERROR_OUTOFMEMORY);
             return FALSE;
         }
-        result = GetCurrentDirectoryW(result, new_path);
-        if (!result) {
+        int retry_result = GetCurrentDirectoryW(result, new_path);
+        if (!retry_result || retry_result > result) {
```

Comment:
Changing the current directory shouldn't fail because of a race condition. I suggest using a while loop here (untested):

    static BOOL __stdcall
    win32_wchdir(LPCWSTR path)
    {
        wchar_t path_buf[MAXPATHLEN], *new_path = path_buf;
        int result, length = Py_ARRAY_LENGTH(path_buf);
        wchar_t env[4] = L"=x:";

        if(!SetCurrentDirectoryW(path))
            return FALSE;

        while (1) {
            wchar_t *temp;
            result = GetCurrentDirectoryW(length, new_path);
            if (!result)
                goto exit;
            if (result <= length)
                break;
            length = result;
            temp = PyMem_RawRealloc(new_path != path_buf ? new_path : NULL,
                length * sizeof(wchar_t));
            if (!temp)
                goto exit;
            new_path = temp;
        }
        
        if (wcsncmp(new_path, L"\\\\", 2) == 0 ||
            wcsncmp(new_path, L"//", 2) == 0)
            /* UNC path, nothing to do. */
            goto exit;
        
        env[1] = new_path[0];
        result = SetEnvironmentVariableW(env, new_path);

    exit:
        if (new_path != path_buf)
            PyMem_RawFree(new_path);
        return result;
    }

The single exit point half ensures the buffer is always freed. I also switched to using `MAXPATHLEN` instead of WinAPI `MAX_PATH`, in case we decide to increase its size later on.

Reply:
Thank you for your suggestion! I've considered using a loop here as in `win32_wgetcwd()`, but deemed `win32_wchdir()` too racy anyway: it may set a stale per-drive current directory in the environment, so I focused on fixing just the crash. But I reconsidered it given your comment and the opportunity to unify the code getting cwd. I've used your loop -- I like it.